wavesexchange_repos/src/circuit_breaker/mod.rs

```diff
@@ -0,0 +1,299 @@
+    max_timespan: Option<Duration>,
+    max_err_count_per_timespan: Option<usize>,
+    init_fn: Option<Box<dyn DataSrcInitFn<S>>>,
+}
+
+impl<S: FallibleDataSource> CircuitBreakerBuilder<S> {
+    pub fn new() -> CircuitBreakerBuilder<S> {
+        CircuitBreakerBuilder {
+            max_timespan: None,
+            max_err_count_per_timespan: None,
+            init_fn: None,
+        }
+    }
+
+    pub fn with_max_timespan(mut self, ts: Duration) -> CircuitBreakerBuilder<S> {
+        self.max_timespan = Some(ts);
+        self
+    }
+
+    pub fn with_max_err_count_per_timespan(mut self, count: usize) -> CircuitBreakerBuilder<S> {
+        self.max_err_count_per_timespan = Some(count);
+        self
+    }
+
+    pub fn with_init_fn(mut self, f: impl DataSrcInitFn<S>) -> CircuitBreakerBuilder<S> {
+        self.init_fn = Some(Box::new(f));
+        self
+    }
+
+    /// Build the circuit breaker.
+    /// Note: you should set up all 3 fields within this builder, either it will panic
+    pub fn build(self) -> Result<CircuitBreaker<S>, S::Error> {
+        // probably there is a better way to force use all with_* methods on builder
+        if self.max_err_count_per_timespan.is_none() {
+            panic!("max_err_count_per_timespan is not set");
+        }
+
+        if self.max_timespan.is_none() {
+            panic!("max_timespan is not set");
+        }
+
+        if self.init_fn.is_none() {
+            panic!("init_fn is not set");
+        }
+
+        let init_fn = self.init_fn.unwrap();
+
+        Ok(CircuitBreaker {
+            state: RwLock::new(CBState {
+                data_source: Arc::new(init_fn()?),
+                err_count: 0,
+                first_err_ts: None,
+            }),
+            max_timespan: self.max_timespan.unwrap(),
+            max_err_count_per_timespan: self.max_err_count_per_timespan.unwrap(),
+            init_fn,
+        })
+    }
+}
+
+impl<S: FallibleDataSource> CircuitBreaker<S> {
+    pub fn builder() -> CircuitBreakerBuilder<S> {
+        CircuitBreakerBuilder::new()
+    }
+
+    pub fn builder_from_cfg(cfg: &Config) -> CircuitBreakerBuilder<S> {
+        Self::builder()
+            .with_max_err_count_per_timespan(cfg.max_err_count_per_timespan)
+            .with_max_timespan(cfg.max_timespan)
+    }
+
+    /// Query the data source. If succeeded, CB resets internal error counter.
+    /// If error returned, counter increases.
+    /// If (N > max_err_count_per_timespan) errors appeared, CB is falling back (panic as default).
+    /// If not enough errors in a timespan appeared to trigger CB's fallback, error counter will be reset.
+    pub async fn query<T, F, Fut>(&self, query_fn: F) -> Result<T, S::Error>
+    where
+        F: FnOnce(Arc<S>) -> Fut,
+        Fut: Future<Output = Result<T, S::Error>>,
+    {
+        let state_read_lock = self.state.read().await;
+        let result = query_fn(state_read_lock.data_source.clone()).await;
```

Comment:
Calling (and awaiting) an external async fn while holding lock makes this `query()` effectively single-threaded which is unacceptable in a real-world application.

Reply:
docs:
An RwLock will allow any number of readers to acquire the lock as long as a writer is not holding the lock.